src/doc/rustc/src/codegen-options/index.md

```diff
@@ -367,12 +401,35 @@ machine. Each target has a default base CPU.
 
 ## target-feature
 
-Individual targets will support different features; this flag lets you control
-enabling or disabling a feature. Each feature should be prefixed with a `+` to
-enable it or `-` to disable it. Separate multiple features with commas.
+This option controls either CPU instruction set extensions that can be used during code generation,
+or static linking.
+
+In both cases `-C target-feature=+x` enables feature `x` and `-C target-feature=-x` disables it.
+
+Multiple features can be specified by separating them with commas - `-C target-feature=+x,-y`.
 
-To see the valid options and an example of use, run `rustc --print
-target-features`.
+For each supported feature `x` `target_feature = "x"` becomes a `cfg` predicate usable early during
+compilation.  
+Code configured with `cfg(target_feature = "x")` is kept if the feature `x` is enabled,
+and removed otherwise.
+
+If the enabled feature is not supported for the current target, the behavior is unspecified.
+<!-- There are multiple bugs here
+  - Unknown CPU features go straight to LLVM and appear as LLVM warnings.
+  - At configure time `crt-static` predicate is set even if the target doesn't support `crt-static`.
+  - At link time `crt-static` is sometimes ignored
+    if the target doesn't support it and sometimes not.
```

Comment:
Also, I strongly suspect that https://github.com/rust-lang/rust/pull/69519 was a hack rather than a proper solution, but I need some time to prepare a musl setup and verify things. (Proc-macros shouldn't be any different to cdylibs and dylibs with regards to linking behavior.)

Reply:
> Proc-macros shouldn't be any different to cdylibs and dylibs with regards to linking behavior

That was the problem 😉

TL;DR It's a hack.

Musl target is a "weirdo". Because of the legacy decisions it defaults to static linking so it cannot produce dylibs, cdylibs and proc-macros. Now lack of proc-macro support doesn't matter when cross-compiling from the host which supports them to musl.
It becomes difficult when using musl as the host toolchain. Since it defaults to static linking proc-macros don't work. To build them one had to use `target-feature=-crt-static` and lose ability to statically link binaries.

Reply:
@mati865 
First, musl toolchain actually supports statically linking self-contained `.so`s, despite the flag in `rustc` target spec saying otherwise, I tried it (at least with "hello world"-like examples).

Second, that doesn't mean that we *want* to use it for proc macros (as well as cdylibs and dylibs) at the same time as we are building a statically linked executable.

One common `RUSTFLAGS="-Ctarget-feature=+crt-static"` (or a target spec default) in this case brings collateral damage on the crates which it is not supposed to apply to.

I see multiple possible solutions here some of which could be more principled than what https://github.com/rust-lang/rust/pull/69519 did.
- Automatically enable `-Ctarget-feature=-crt-static` for proc macro crates or all libraries in Cargo, modifying RUSTFLAGS. Question: how to opt-out?
- Introduce a new options `-Ctarget-feature=+crt-static-dylib` controlling static linking of libraries instead of `-Ctarget-feature=-crt-static`. It would almost never be used on Linux (not sure about windows-msvc and wasm).
- Keep the existing meaning of `crt-static`, but introduce a new option `-C disable-crt-static-for-dylibs` or something. Cargo would then use it for proc macro crates or all libraries. Question: how to opt-out?
- Ignore `+crt-static` for dylibs if the target doesn't support it. This is fragile, musl actually supports it despite the current value of the flag of the musl target spec. If the flag is enabled, proc macros will break.
- Ignore `+crt-static` for dylibs always. They are almost never used on Linux (not sure about windows-msvc and wasm). `+crt-static-dylib` looks strictly better.

Reply:
- Have two `crt-static` defaults in target specs - one for executables and one for libraries. Solves one half of the problem, but explicit `+crt-static` in RUSTFLAGS will still cause collateral damage.

Reply:
I'll turn this into a proper issue.

Reply:
https://github.com/rust-lang/rust/issues/71651